### ISSUE TITLE
Git branch left with detached head after run.

### ISSUE THREAD

**Issue by GrabCAD-Gary** [megalinter/megalinter#810]
**Describe the bug**
Linting regualrly fails with:

```
Pulling docker image nvuillam/mega-linter-dotnet:v4 ...
INFO: this operation can be long during the first use of mega-linter-runner
The next runs, it will be immediate (thanks to docker cache !)
v4: Pulling from nvuillam/mega-linter-dotnet
Digest: sha256:30cc117eeb229e86b963e...
Status: Image is up to date for nvuillam/mega-linter-dotnet:v4
docker.io/nvuillam/mega-linter-dotnet:v4
Command: docker run -v /var/run/docker.sock:/var/run/docker.sock:rw -v C:\Code\<path>:/tmp/lint:rw -e ENABLE_LINTERS=POWERSHELL_POWERSHELL -e EXCLUDED_DIRECTORIES=<dirs> nvuillam/mega-linter-dotnet:v4
[config] /tmp/lint/.mega-linter.yml + Environment variables
----------------------------------------------------------------------------------------------------
------------------------------------------- Mega-Linter --------------------------------------------
----------------------------------------------------------------------------------------------------
 - Image Creation Date: 2021-09-21T20:49:26Z
 - Image Revision: 96137434
 - Image Version: v4
----------------------------------------------------------------------------------------------------
The Mega-Linter documentation can be found at:
 - https://nvuillam.github.io/mega-linter
----------------------------------------------------------------------------------------------------
GITHUB_REPOSITORY:
GITHUB_REF:
GITHUB_RUN_ID:


To receive reports as email, please set variable EMAIL_REPORTER_EMAIL
Listing updated files in [/tmp/lint] using git diff, then filter with:
Warning: Unable to pull current branch 2b51a6371ff...
>> Traceback (most recent call last):
>>   File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
>>     return _run_code(code, main_globals, None,
>>   File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
>> exec(code, run_globals)
>>   File "/megalinter/run.py", line 15, in <module>
>>     linter.run()
>>   File "/megalinter/MegaLinter.py", line 105, in run
>>     self.collect_files()
>>   File "/megalinter/MegaLinter.py", line 386, in collect_files
>>     all_files = self.list_files_git_diff()
>>   File "/megalinter/MegaLinter.py", line 493, in list_files_git_diff
>>     repo.git.checkout(default_branch)
>>   File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 638, in <lambda>
>> return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
>>   File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 1183, in _call_process
>> return self.execute(call, **exec_kwargs)
>>   File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 983, in execute
>> raise GitCommandError(redacted_command, status, stderr_value, stdout_value)
>> git.exc.GitCommandError: Cmd('git') failed due to: exit code(1)
>>   cmdline: git checkout master
>>   stderr: 'error: Your local changes to the following files would be overwritten by checkout:
>>      .mega-linter.yml
>>      grunt-tasks/lint.js
>> Please commit your changes or stash them before you switch branches.
>> Aborting'
Warning: Command failed: npx mega-linter-runner --flavor=dotnet -e 'ENABLE_LINTERS=POWERSHELL_POWERSHELL' -e 'EXCLUDED_DIRECTORIES=desktop-platform-installer,desktop,scripts,geometry-core'
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/megalinter/run.py", line 15, in <module>
    linter.run()
  File "/megalinter/MegaLinter.py", line 105, in run
    self.collect_files()
  File "/megalinter/MegaLinter.py", line 386, in collect_files
    all_files = self.list_files_git_diff()
  File "/megalinter/MegaLinter.py", line 493, in list_files_git_diff
    repo.git.checkout(default_branch)
  File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 638, in <lambda>
    return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 1183, in _call_process
    return self.execute(call, **exec_kwargs)
  File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 983, in execute
    raise GitCommandError(redacted_command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(1)
  cmdline: git checkout master
  stderr: 'error: Your local changes to the following files would be overwritten by checkout:
        .mega-linter.yml
        grunt-tasks/lint.js
Please commit your changes or stash them before you switch branches.
Aborting'
 Use --force to continue.

Aborted due to warnings.
```

This leaves my branch with a detached head.  That's not acceptable. If there is some git-related problem the process should exit gracefully without damaging the environment. Why it thinks it's changing branches is a mystery. I assumed it to be related to git-diff operations (`ENABLE: GIT`) but disabling that didn't stop it. git_diff is reported as running, though. Updating from master has appeared to help on previous occasions but didn't on the last such occurrence. My best guess is that somwhere internal it's trying to check out branches to identify changes, even though`VALIDATE_ALL_CODEBASE: false` is set, and something in the temp location is somehow showing changes that it can't overwrite. 

***Work-around***
I can recover by stashing changes, checking out and re-applying the stash

**To Reproduce**
Unclear. It works for most runs and then starts failing this way for consecutive runs until something somewhere recovers. 
However, see the additional context below. I'd need to test further to confirm whether that's the trigger.

**Expected behavior**
Internal diff management, if that's what's failing, should work reliably. Failures shouldn't leave branches in broken states.

**Screenshots**

**Additional context**

Disabling `VALIDATE_ALL_CODEBASE: false` appears to have stopped the error occurring so it seems that it's related to that setting.

**Comment by GrabCAD-Gary** [comment]
The culprit seems confirmed as `VALIDATE_ALL_CODEBASE: false`.  Sequence:

Enable `VALIDATE_ALL_CODEBASE: false` in mega_linter.yml and run. Error occurs.
Comment out `VALIDATE_ALL_CODEBASE: false` and run. Linting succeeds.
Uncomment `VALIDATE_ALL_CODEBASE: false` and run. Error occurs.
Change entry to `VALIDATE_ALL_CODEBASE: true`. Linting succeeds.

**Comment by noskovao** [comment]
@GrabCAD-Gary try to use the latest docker image since #760 got merged

**Comment by GrabCAD-Gary** [comment]
Doesn't work, I'm afraid. With `npx mega-linter-runner --flavor=dotnet -r insiders ...` it fails at the outset with:
```
...
To receive reports as email, please set variable EMAIL_REPORTER_EMAIL
Listing updated files in [/tmp/lint] using git diff.
>> Traceback (most recent call last):
>>   File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
>> return _run_code(code, main_globals, None,
>>   File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
>>     exec(code, run_globals)
>>   File "/megalinter/run.py", line 15, in <module>
>>     linter.run()
>>   File "/megalinter/MegaLinter.py", line 105, in run
>>     self.collect_files()
>>   File "/megalinter/MegaLinter.py", line 386, in collect_files
>>     all_files = self.list_files_git_diff()
>>   File "/megalinter/MegaLinter.py", line 483, in list_files_git_diff
>>     repo.git.fetch()
>>   File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 638, in <lambda>
>> return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
>>   File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 1183, in _call_process
>> return self.execute(call, **exec_kwargs)
>>   File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 983, in execute
>> raise GitCommandError(redacted_command, status, stderr_value, stdout_value)
>> git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
>>   cmdline: git fetch
>>   stderr: 'Host key verification failed.
>> fatal: Could not read from remote repository.
>>
>> Please make sure you have the correct access rights
>> and the repository exists.'
Warning: Command failed: npx mega-linter-runner --flavor=dotnet -r insiders -e 'ENABLE_LINTERS=POWERSHELL_POWERSHELL'
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/megalinter/run.py", line 15, in <module>
    linter.run()
  File "/megalinter/MegaLinter.py", line 105, in run
    self.collect_files()
  File "/megalinter/MegaLinter.py", line 386, in collect_files
    all_files = self.list_files_git_diff()
  File "/megalinter/MegaLinter.py", line 483, in list_files_git_diff
    repo.git.fetch()
  File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 638, in <lambda>
    return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 1183, in _call_process
    return self.execute(call, **exec_kwargs)
  File "/usr/local/lib/python3.9/site-packages/git/cmd.py", line 983, in execute
    raise GitCommandError(redacted_command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git fetch
  stderr: 'Host key verification failed.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.'
```

**Comment by GrabCAD-Gary** [comment]
I ran again with today's latest but it failed the same way with `VALIDATE_ALL_CODEBASE: false` set. It seems to me that it's trying to run git-diff to determine what files to check. I've been working in the repo outside of M-L continuously without issue, including running `git diff`. Furthermore, the Git 'linter' doing the same thing runs happily. Can you explain further why you think that there is some action required on my side and what you're suggesting I need to do? Thanks.

**Comment by Kurt-von-Laven** [comment]
Fixed in v5.0.0.
